retrieval/discovery/kubernetes/discovery.go

```diff
@@ -774,3 +816,206 @@ func nodeHostIP(node *Node) (net.IP, error) {
+		log.Debugf("Got pod %s", pod.ObjectMeta.Name)
+		podMap[pod.ObjectMeta.Name] = &pods.Items[idx]
+	}
+
+	return podMap, pods.ResourceVersion, nil
+}
+
+func (kd *Discovery) watchPods(events chan interface{}, done <-chan struct{}, retryInterval time.Duration) {
+	until(func() {
+		pods, resourceVersion, err := kd.getPods()
+		if err != nil {
+			log.Errorf("Cannot initialize pods collection: %s", err)
+			return
+		}
+
+		kd.pods = map[string]map[string]*Pod{}
+
+		for _, pod := range pods {
+			events <- &podEvent{added, pod}
+		}
+
+		req, err := http.NewRequest("GET", podsURL, nil)
+		if err != nil {
+			log.Errorf("Cannot create pods request: %s", err)
+			return
+		}
+
+		values := req.URL.Query()
+		values.Add("watch", "true")
+		values.Add("resourceVersion", resourceVersion)
+		req.URL.RawQuery = values.Encode()
+		res, err := kd.queryAPIServerReq(req)
+		if err != nil {
+			log.Errorf("Failed to watch pods: %s", err)
+			return
+		}
+		defer res.Body.Close()
+		if res.StatusCode != http.StatusOK {
+			log.Errorf("Failed to watch pods: %d", res.StatusCode)
+			return
+		}
+
+		d := json.NewDecoder(res.Body)
+
+		for {
+			var event podEvent
+			if err := d.Decode(&event); err != nil {
+				log.Errorf("Watch pods unexpectedly closed: %s", err)
+				return
+			}
+
+			select {
+			case events <- &event:
+			case <-done:
+			}
+		}
+	}, retryInterval, done)
+}
+
+func podSource(pod *Pod) string {
+	return sourcePodPrefix + ":" + pod.ObjectMeta.Name
+}
+
+func updatePodTargets(pod *Pod, allContainers bool) []model.LabelSet {
+	var targets []model.LabelSet = make([]model.LabelSet, 0, len(pod.PodSpec.Containers))
+	if pod.PodStatus.PodIP == "" {
+		log.Debugf("skipping pod %s -- PodStatus.PodIP is empty", pod.ObjectMeta.Name)
+		return targets
+	}
+
+	if pod.PodStatus.Phase != "Running" {
+		log.Debugf("skipping pod %s -- status is not `Running`", pod.ObjectMeta.Name)
+		return targets
+	}
+
+	for _, container := range pod.PodSpec.Containers {
+	PortLoop:
+		for _, port := range container.Ports {
+			if port.Protocol == "TCP" {
+				t := model.LabelSet{
+					model.AddressLabel:        model.LabelValue(net.JoinHostPort(pod.PodIP, strconv.Itoa(port.ContainerPort))),
+					podNameLabel:              model.LabelValue(pod.ObjectMeta.Name),
+					podAddressLabel:           model.LabelValue(pod.PodStatus.PodIP),
+					podNamespaceLabel:         model.LabelValue(pod.ObjectMeta.Namespace),
+					podContainerNameLabel:     model.LabelValue(container.Name),
+					podContainerPortNameLabel: model.LabelValue(port.Name),
+				}
+
+				for k, v := range pod.ObjectMeta.Labels {
+					labelName := strutil.SanitizeLabelName(podLabelPrefix + k)
+					t[model.LabelName(labelName)] = model.LabelValue(v)
+				}
+
+				for k, v := range pod.ObjectMeta.Annotations {
+					labelName := strutil.SanitizeLabelName(podAnnotationPrefix + k)
+					t[model.LabelName(labelName)] = model.LabelValue(v)
+				}
+
+				targets = append(targets, t)
+				break PortLoop
```

Comment:
We're only want at most one port returned per container, so we need to choose one deterministically and allow relabelling to get at the rest.

That'd change this from pod to container SD.